### PR TITLE
Fixed #11385 -- Made forms.DateTimeField accept ISO 8601 date inputs.

### PR DESCRIPTION
**Pull request by claudep**
Thanks José Padilla for the initial patch.

### CODE REVIEW

File: django/utils/formats.py [hunk 1]
```diff
@@ -23,7 +23,11 @@
         '%Y-%m-%d %H:%M:%S',
         '%Y-%m-%d %H:%M:%S.%f',
         '%Y-%m-%d %H:%M',
-        '%Y-%m-%d'
+        '%Y-%m-%d',
+        # ISO 8601
+        '%Y-%m-%dT%H:%M:%S',
+        '%Y-%m-%dT%H:%M:%S.%f',
+        '%Y-%m-%dT%H:%M',
```
Comment (carltongibson) [inline django/utils/formats.py]: Can we add support for the [offset and `Z` formats](https://en.wikipedia.org/wiki/ISO_8601#Time_offsets_from_UTC) too? (They're used **a lot**.)

File: docs/releases/3.1.txt [hunk 2]
```diff
@@ -130,7 +130,7 @@ File Uploads
 Forms
 ~~~~~
 
-* ...
+* :attr:`django.forms.DateTimeField` accepts dates in ISO 8601 format.
```
Comment (carltongibson) [inline docs/releases/3.1.txt]: I think this is a major feature. Maybe it should go near the top? 🙂

If not, perhaps expanding a little bit...

File: django/utils/formats.py (hunk 1 above)
Comment (claudep) [reply django/utils/formats.py]: I worked to add that support, but then stumbled on timezones.tests.NewFormsTests.test_form_with_explicit_timezone (https://code.djangoproject.com/ticket/19371). So should we or should we not accept timezones in DateTimeField inputs?

**Comment by aaugustin** [comment]
IIRC the reasons for not accepting time zone information in date time form fields were:

- Forms are intended for user input (remember, this was 2011). Users don't tend to input time zone information spontaneously. When we decided to support ISO 8601, we made this argument irrelevant.
- Naive values should be delineated clearly from aware values. Forms are the frontier that converts between naive and aware values. Fields in HTML pages are in "naive values" land. This is a fairly theoretical argument. It may not hold if you counter it with "be strict in what you generate, lax in what you accept".
- I didn't want to add complexity that didn't have an obvious use case. Apparently we have a use case now.

None of these arguments should block progress with this ticket. Just generate a suitable `datetime.timezone`.

**Comment by carltongibson** [comment]
OK, thanks @aaugustin. Let's support offsets and `Z`.

File: docs/ref/forms/fields.txt [hunk 3]
```diff
@@ -506,20 +506,30 @@ For each field, we describe the default widget used if you don't specify
...
-         '%Y-%m-%d %H:%M',       # '2006-10-25 14:30'
-         '%Y-%m-%d',             # '2006-10-25'
-         '%m/%d/%Y %H:%M:%S',    # '10/25/2006 14:30:59'
-         '%m/%d/%Y %H:%M',       # '10/25/2006 14:30'
-         '%m/%d/%Y',             # '10/25/2006'
-         '%m/%d/%y %H:%M:%S',    # '10/25/06 14:30:59'
-         '%m/%d/%y %H:%M',       # '10/25/06 14:30'
-         '%m/%d/%y']             # '10/25/06'
+    The field always accept strings in ISO 8601 formatted dates or similar
+    recognized by :func:`django.utils.dateparse.parse_datetime`. Some examples
```
Comment (carltongibson) [inline docs/ref/forms/fields.txt]: So, question, are we sure about _always_? (i.e. Do we want to allow disabling? — Perhaps we can do that IF someone actually wants it...)

Need as `s` on `accepts`. 

Can I suggest an addition to the `attribute` description above too: 

```
A list of formats used to attempt to convert a string to a valid 
``datetime.datetime`` object, in addition to ISO 8601 formats. 
```

Not 100% happy about how that reads, which might be a pointer towards the extra API... (???)

Comment (claudep) [reply docs/ref/forms/fields.txt]: I don't see currently any use case for *not* accepting ISO strings. If such a case would arise, then you can always use a customize `to_python` method.

Comment (carltongibson) [reply docs/ref/forms/fields.txt]: Ok, me neither really. Maybe one appears but... DEAL_WITH_LATER.

**Comment by claudep** [comment]
@carltongibson, still in your review queue?

File: docs/ref/forms/fields.txt [hunk 4]
```diff
@@ -504,22 +504,32 @@ For each field, we describe the default widget used if you don't specify
...
-         '%m/%d/%Y %H:%M:%S',    # '10/25/2006 14:30:59'
-         '%m/%d/%Y %H:%M',       # '10/25/2006 14:30'
-         '%m/%d/%Y',             # '10/25/2006'
-         '%m/%d/%y %H:%M:%S',    # '10/25/06 14:30:59'
-         '%m/%d/%y %H:%M',       # '10/25/06 14:30'
-         '%m/%d/%y']             # '10/25/06'
+    If no ``input_formats`` argument is provided, the default input formats are
+    taken from :setting:`DATETIME_INPUT_FORMATS` if :setting:`USE_L10N` is
+    ``False``, or from the active locale format ``DATETIME_INPUT_FORMATS`` key
+    if localization is enabled.
```
Comment (felixxm) [inline docs/ref/forms/fields.txt]: This small clarification is valid also without this change so we can move it to a separate commit.

**Comment by felixxm** [comment]
@claudep Thanks :+1: I corrected notes about the default input formats also for `DateField` and `TimeField`, added tests for stripping whitespaces in ISO formats, pushed minor edits, and rebased from master.

**Comment by claudep** [comment]
Awesome! You could even have set you as commit author, as my contributions are now minor :-)
